File: leetcode/221_maximal_square.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
    int maximalSquare(std::vector<std::vector<char>>& matrix) {
        if (matrix.empty()) {
            return 0;
        }

        int m = static_cast<int>(matrix.size());
        int n = static_cast<int>(matrix[0].size());
        int max_side_length = 0;
        std::vector<int> dp(n, 0);

        for (int i = m - 1; i >= 0; --i) {
            int diag = 0;
            for (int j = 0; j < n; ++j) {
                int left = (j - 1 < 0) ? 0 : dp[j - 1];
                int bottom = dp[j];
                if (matrix[i][j] == '1') {
                    // We are consistently limited by the smallest of the 3: left, bottom, diag
                    int side_length = std::min({left, bottom, diag});
                    dp[j] = side_length + 1;
                } else {
                    dp[j] = 0;
                }
                diag = bottom;
                max_side_length = std::max(max_side_length, dp[j]);
            }
        }
        return max_side_length * max_side_length;
    }
};

File: leetcode/221_maximal_square.py
from typing import List


class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        # A square of size n + 1 can be formed with top left corner at dp[i][j] if dp[i + 1][j], dp[i + 1][j + 1], and dp[i][j + 1] are of size n
        # and matrix[i][j] is 1

        largest = 0
        m = len(matrix)
        n = len(matrix[0])
        dp = [0 for _ in range(n)] # start with state of last row, iterate right->left bottom->top
        for i in range(m - 1, -1, -1):
            right = 0
            diag = 0
            for j in range(n - 1, -1, -1):
                bottom = dp[j]
                potential = min(right, diag, bottom) + 1
                if matrix[i][j] == '1':
                    dp[j] = potential
                    largest = max(largest, potential)
                else:
                    dp[j] = 0

                diag = bottom
                right = dp[j]

        return largest * largest
